src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+import { debounce, DebouncedFunc } from './utils';
+import { StateStore } from './store';
+import type { Channel } from './channel';
+import type { StreamChat } from './client';
+import type {
+  ChannelFilters,
+  ChannelOptions,
+  ChannelSort,
+  DefaultGenerics,
+  ExtendableGenerics,
+  MessageFilters,
+  MessageResponse,
+  SearchMessageSort,
+  SearchOptions,
+  UserFilters,
+  UserOptions,
+  UserResponse,
+  UserSort,
+} from './types';
+
+export type SearchSourceType = 'channels' | 'users' | 'messages' | (string & {});
+export type QueryReturnValue<T> = { items: T[]; next?: string };
+export type DebounceOptions = {
+  debounceMs: number;
+};
+type DebouncedExecQueryFunction = DebouncedFunc<(searchString?: string) => Promise<void>>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export interface SearchSource<T = any> {
+  activate(sourceStateOverride?: Partial<SearchSourceState>): void;
+  deactivate(): void;
+  readonly hasMore: boolean;
+  readonly hasResults: boolean;
+  readonly isActive: boolean;
+  readonly isLoading: boolean;
+  readonly items: T[] | undefined;
+  readonly lastQueryError: Error | undefined;
+  readonly next: string | undefined;
+  readonly offset: number | undefined;
+  resetState(stateOverrides?: Partial<SearchSourceState<T>>): void;
+  search(text?: string): void;
+  searchDebounced: DebouncedExecQueryFunction;
+  readonly searchQuery: string;
+  setDebounceOptions(options: DebounceOptions): void;
+  readonly state: StateStore<SearchSourceState<T>>;
+  readonly type: SearchSourceType;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SearchSourceState<T = any> = {
+  hasMore: boolean;
+  isActive: boolean;
+  isLoading: boolean;
+  items: T[] | undefined;
```

Comment:
```suggestion
  items: T[] | undefined;
```
Why `undefined`? Can't it be an empty array instead?

Reply:
The value `undefined` says that no search has been executed yet. This is handy to know if we want to differ btw the state where will tell the user to start search vs telling the user there are not results for the given search query.

We could think that maybe the same could be inferred from the combination of the length of the search query and the length of the items array. The problem is, when you start typing, for a split of a second, the value of loading is not set. In that case you have a search query of non-zero lenght, loading is false and results are an empty array. That leads to showing "No results found" or whatever the empty results component is. So I would say we need an indicator that tells us, whether the items value is the result of a search query.